sql/src/main/java/io/crate/operation/fetch/LuceneDocFetcher.java

```diff
@@ -0,0 +1,139 @@
+                            List<LuceneCollectorExpression<?>> collectorExpressions,
+                            RowDownstream downstream,
+                            NodeFetchOperation.ShardDocIdsBucket shardDocIdsBucket,
+                            JobCollectContext jobCollectContext,
+                            CrateSearchContext searchContext,
+                            int jobSearchContextId,
+                            boolean closeContext) {
+        inputRow = new InputRow(inputs);
+        this.collectorExpressions = collectorExpressions;
+        this.downstream = downstream.registerUpstream(this);
+        this.shardDocIdsBucket = shardDocIdsBucket;
+        this.jobCollectContext = jobCollectContext;
+        this.searchContext = searchContext;
+        this.jobSearchContextId = jobSearchContextId;
+        this.closeContext = closeContext;
+        this.fieldsVisitor = new LuceneDocCollector.CollectorFieldsVisitor(collectorExpressions.size());
+    }
+
+    public void setNextReader(AtomicReaderContext context) throws IOException {
+        currentReader = context.reader();
+        for (LuceneCollectorExpression expr : collectorExpressions) {
+            expr.setNextReader(context);
+        }
+    }
+
+    private void fetch(long position, int doc) throws Exception {
+        if (ramAccountingContext != null && ramAccountingContext.trippedBreaker()) {
+            // stop fetching because breaker limit was reached
+            throw new UnexpectedFetchTerminatedException(
+                    CrateCircuitBreakerService.breakingExceptionMessage(ramAccountingContext.contextId(),
+                            ramAccountingContext.limit()));
+        }
+        if (visitorEnabled) {
+            fieldsVisitor.reset();
+            currentReader.document(doc, fieldsVisitor);
+        }
+        for (LuceneCollectorExpression e : collectorExpressions) {
+            e.setNextDocId(doc);
+        }
+        if (!downstream.setNextRow(new PositionalRowDelegate(inputRow, position))) {
+            // no more rows required, we can stop here
+            throw new FetchAbortedException();
+        }
+
+    }
+
+    public void doFetch(RamAccountingContext ramAccountingContext) throws Exception {
+        this.ramAccountingContext = ramAccountingContext;
+
+        jobCollectContext.acquireContext(searchContext);
+
+        CollectorContext collectorContext = new CollectorContext()
+                .visitor(fieldsVisitor)
+                .searchContext(searchContext)
+                .searchLookup(searchContext.lookup(false));
+        for (LuceneCollectorExpression<?> collectorExpression : collectorExpressions) {
+            collectorExpression.startCollect(collectorContext);
+        }
+        visitorEnabled = fieldsVisitor.required();
+
+        try {
+            for (int index = 0; index < shardDocIdsBucket.size(); index++) {
+                int docId = shardDocIdsBucket.docId(index);
```

Comment:
isn't this the same as in the LuceneDocCollector? Is it possible to share something here?